thrift/arg2/kv_iterator.go

```diff
@@ -0,0 +1,87 @@
+package arg2
+
+import "encoding/binary"
+
+// KeyValIterator is a iterator for reading tchannel-thrift Arg2 Scheme,
+// which has key/value pairs (k~2 v~2).
+// NOTE: to be optimized for performance, we try to limit the allocation
+// done in the process of iteration.
+type KeyValIterator struct {
+	arg2Len               int
+	leftPairCount         int
+	keyOffset, keyLen     int
+	valueOffset, valueLen int
```

Comment:
do we need offset + length for both of these? i'd assume `keyOffset` and `valueOffset` are enough to figure out `keyLen`?

Reply:
That's a good optimization.